FLOE-116: Transcripts Panel

jobara opened the pull request:
Large scale refactoring to separate out the reusable pieces from the captions panel. Used this to implement the Transcripts Panel.

cindyli committed:
FLOE-116: Created a base panel grade to be shared by all metadata panels.

cindyli committed:
FLOE-116: Added unit tests for panel component.

cindyli committed:
FLOE-116: Modified the panel unit test to pass in the initial model value for the indicator component.

cindyli committed:
FLOE-116: Properly scoped the panel selectors.

cindyli committed:
FLOE-116: Modified the audio metadata panel to make use of the base panel grade.

jobara committed:
FLOE-116: Using model transformations in panel

The panel base grade uses an updateModel function to transform a given model, with the supplied transformation rules and updates the model of the indicator.

jobara committed:
Merge remote-tracking branch 'cindy/FLOE-116' into FLOE-116

jobara committed:
FLOE-116: renamed rules option

cindyli committed:
FLOE-116: Updated the audio panel component to add in the model transformation rules for the indicator, which is newly introduced to the base panel grade.

cindyli committed:
FLOE-116: Modified the base panel grade to perform the auto-generate the initial indicator model.

cindyli committed:
FLOE-116: Simplified the audio panel, indicator subcomponent.

cindyli committed:
FLOE-116: Split the base panel grade to a basePanel component that has the most fundamental configurations and a Panel grade that brings the base panel grade one step further with auto-generating the indicator model by transforming the panel model.

cindyli committed:
FLOE-116: Modified video and audio panels to make use of improved base panel grades.

jobara committed:
FLOE-116: Added resourceInputPanel

Added a panel grade for resource input.

jobara committed:
Merge remote-tracking branch 'cindy/FLOE-116' into FLOE-116

jobara committed:
Merge branch 'master' into FLOE-116

jobara committed:
FLOE-116: Updated the all-tests.html file

jobara committed:
FLOE-116: Updated captions with resourceInputPanel

jobara committed:
FLOE-116: strings handled by resourceInputPanel

The resouceInputPanel now handles the strings for the sub components, instead of making integrations do this.

jobara committed:
FLOE-116: Moved template loading to base grade.

Moved template loading into the baseResourceInputPanel grade. Made a common template files for both the resourceInputPanel grades.

jobara committed:
FLOE-116: Adding style to container

Can now provide a style to add to the components container. This is useful for assigning the icon font to the indicator. Also fixed a typo.

jobara committed:
FLOE-116: Fixed captionsPanel

captionsPanel updated to take advantage of the changes to resourceInputPanel. Also modified how the template for the resourceInput is passed into the panel. Now it is part of the resources block, instead of its own option.

jobara committed:
FLOE-116: fixing copyrights

jobara committed:
FLOE-116: Modified class names

Fixed stlying class names used in resourceInputPanel-template to better match its use.

jobara committed:
FLOE-116: added container style to captionsPanel

jobara committed:
FLOE-116: refactored resourceInput strings

Renamed the grouping of strings for the resourceInputs in the resourceInputPanel.

jobara committed:
FLOE-116: Refactored resourceInputPanel tests

Broke out common test grades to be used when testing resourceInputPanel based grades. Also fixed a bug with the panel title.

jobara committed:
FLOE-116: Added transcripts panel

jobara committed:
FLOE-116: Update the all-tests.html file

added the transcripts tests

jobara committed:
Merge branch 'master' into FLOE-116

Conflicts:
	demos/html/metadata.html
	tests/all-tests.html

cindyli committed:
FLOE-116: Fixed the issue that changes made in captions panel are not saved.

jobara committed:
FLOE-116: removed console.log

jobara commented on src/js/panel.js:
```diff
@@ -0,0 +1,94 @@
+    fluid.metadata.panel.updateModel = function (that) {
+        var model = that.transformPanelModel();
+        that.applier.requestChange("value", model.value);
+    };
```
needs a comment to explain the model transformation and relay.

cindyli committed:
FLOE-116: Added comments for model transformation and relay btw panels and indicators.

cindyli committed:
FLOE-116: Fixed the missing audio and video panel titles.

michelled commented on src/js/resourceInputPanel.js:
```diff
@@ -0,0 +1,283 @@
+                    onCreate: "onCreate",
+                    afterMarkupReady: "afterMarkupReady",
+                    afterRenderPrimaryCaption: "afterRenderPrimaryCaption",
+                    afterRenderSecondaryCaption: "afterRenderSecondaryCaption",
```
It looks like there is a stray comma here.

jobara committed:
FLOE-116: Clean up

removed stray comma

jobara commented:
@michelled, I fixed the stray comma, thanks for spotting that.

amb26 commented on src/js/resourceInputPanel.js:
```diff
@@ -0,0 +1,283 @@
+        gradeNames: [ "fluid.multiInputTransformFunction", "fluid.standardOutputTransformFunction" ]
+    });
+
+    fluid.transforms.conditionAny = function (inputs, transformSpec, transform) {
```
Please describe this transform fully, at least with a comment block and preferably with a wiki page, especially if it is going into the top-level fluid.transforms space. What are its inputs, requirements on these inputs, outputs, and any expectations regarding short-circuiting - how does it differ from the existing condition transform that we have?

jobara replied:
I've added a comment to this. Please let me know if it is sufficient. I didn't create a wiki page yet, as I don't know if this will be used elsewhere. I imagine if we did want to move this up to infusion,  we may want to either update the condition transform itself, or generalize this one to handle && as well as the currently implemented ||.

amb26 commented on src/js/resourceInputPanel.js:
```diff
@@ -0,0 +1,283 @@
+
+        var conditions = transformSpec.conditionPath ? fluid.get(transform.source, transformSpec.conditionPath) : transformSpec.condition;
+        var innerPath = transformSpec.innerPath || "";
+        for (var i=0; i <= conditions.length; i++) {
```
Why is this upper loop bound inclusive? It appears to me that it will access an element beyond the end of the array.

amb26 commented on src/js/resourceInputPanel.js:
```diff
@@ -0,0 +1,283 @@
+    });
+
+    fluid.metadata.baseResourceInputPanel.updateModel = function (that, value, path, index, root) {
+        var changePath = [root, index, path].join(".");
```
This join is unnecessary - all ChangeApplier and accessor functions have accepted path segment arrays for some time

jobara replied:
@amb26 when did that change go in. Perhaps the version of Infusion we are using is too old, because this doesn't work using as a straight array. I get a "path.charAt is not a function" error.

amb26 replied:
That fix has been in for at least 6 months (looks like about Sept 2013). It might be worth updating to the latest Infusion in any case to make any subsequent merges easier.

jobara replied:
@amb26, @cindyli is upgrading infusion in her work for FLOE-171, which has merged this branch in already. Is it okay if we modify this in her branch instead?

jobara replied:
@amb26 infusion was updated in master, so we have a very recent copy in this branch now. However, this still didn't work.

``` javascript
that.applier.requestChange([root, index, path], value);
```

causes "path.charAt is not a function" error.

where path is ["resources", 0, ["src"]]

amb26 commented on src/js/videoPanel.js:
```diff
@@ -27,7 +27,18 @@ var fluid_1_5 = fluid_1_5 || {};
+                priority: "first",
+                options: {
+                    model: {
+                        value: "available"
```
This default value appears to duplicate the capabilities of the transform block earlier (sent to "conditionAny"). The "indicator" component needs much better documentation regarding the expectations on its model contents. It should probably also be reorganised so that it is capable of making use of a string bundle - it's unlikely that integrators will discover how to configure its "tooltipContent" block. I notice that references to a string bundle are used elsewhere that the indicator is used, but not here - this should be made consistent.

jobara replied:
The model value is set here to "available" because for the video panel it is always in the available state. It is never changed and doesn't undergo any model transformations. I do agree that the indicator should use the strings block for the tooltip strings.

amb26 commented on src/js/resourceInputPanel.js:
```diff
@@ -0,0 +1,283 @@
+            }]
+        },
+        components: {
+            //TODO: use dynamic components instead of hard coding two resources
```
Yes, please do fix this before merging - we need to keep this cut n pasted component block out of trunk

jobara replied:
@amb26  and I spoke about how to address this today. The suggested approach is to use event driven dynamic components. On the initial creation of the parent component we'll loop through the model array and fire the event for each captions object found. We'll either need a pair of events, or to order the operation so that we inject the appropriate markup fragment into the DOM ahead of the dynamic component being instantiated. 

To support deletion we'll need to have each of the dynamic components fire an event on it's instantiation to let the parent know what piece of the model it refers to and a reference back to itself so that the destroy method can be called on it. Note that if we remove items from the model's array of caption objects we'll need to update the references to all of the dynamic components as well. Maybe this can be simplified by having parallel arrays so that we can just remove the same location in both arrays.

jobara committed:
FLOE-116: Updated conditionAny transform

Added a detailed comment describing its usage. Also fixed an issues with the for loop.

jobara committed:
Merge branch 'master' into FLOE-116

cindyli committed:
FLOE-116: Move the declaration of tooltip display strings into the "strings" option block.

jobara committed:
FLOE-116: Switched to using dynamic components.

The resourceInputs are now added as dynamic components. This is just feature complete with what had been there. No new features to add/remove resourceInputs were added yet.The unit tests were updated to work with the dynamic components.

jobara committed:
Merge remote-tracking branch 'cindy/FLOE-116' into FLOE-116

Conflicts:
	src/js/resourceInputPanel.js

jobara commented:
@amb26 this pull request is ready for another round of review.

amb26 commented on src/js/resourceInputPanel.js:
```diff
@@ -0,0 +1,348 @@
+            }
+        },
+        listeners: {
+            "onCreate.init": "fluid.metadata.resourceInput.init"
```
This isn't a great name or namespace for this listener - it should reflect the actual function of the particular listener, and so should be something like "resources" or "fetchResources"

amb26 commented on src/js/resourceInputPanel.js:
```diff
@@ -0,0 +1,348 @@
+     ************************************/
+
+    /*
+     * The conditionAny transform is similar to that of the standard condition transform found in Infusion.
```
Thanks for this comment and description of the transformer - this makes its purpose enormously more clear. Now I think I understand what it is for, it seems that its name could be a bit more accurate. It seems that the only difference is that this transformer uses a supplied path, "innerPath" for indirection into supplied objects, which must take the form of an array. This suggests a name like "fluid.transforms.conditionWithPath" or even "fluid.transforms.find" since the contract is now extremely similar to our existing "fluid.find" utility

amb26 commented on src/js/resourceInputPanel.js:
```diff
@@ -0,0 +1,348 @@
+                        }
+                    },
+                    modelListeners: {
+                        "*": {
```
Ensure that you really require "*" here rather than "" - cf. recent experiences with @cindyli - "" implies you will receive a single notification for each change whereas with "*" you will receive multiple notifications for changes with different top-level path segments

jobara replied:
Does this also refer to a new change that has happened in the framework within the last 6 months. It doesn't seem to work with "".

jobara replied:
Changing `"*"` to `""` breaks the resourceInputPanel both in the tests and the demos. I've left it at `"*"`.

amb26 replied:
Ok - this is because the entire hierarchy is still using the "old ChangeApplier" implementation. Since this is new code we should make sure to migrate it over to the new hierarchy involving "modelRelayComponent", "viewRelayComponent" etc. I believe cindy has a branch where this is done? If this is being kept up to date, we should just leave a note to make sure that these changes should be ported over there.

jobara replied:
I've left a comment about making this change.

amb26 commented on src/js/panel.js:
```diff
@@ -0,0 +1,97 @@
+    // panel model to the indicator model.
+    fluid.metadata.panel.updateModel = function (that) {
+        var model = that.transformPanelModel();
+        that.applier.requestChange("value", model.value);
```
If you are up-to-date with the latest Infusion, it is probably time to start switching to the new standard of using "change" rather than "requestChange". We will deprecate "requestChange" in the 1.5 release and eliminate it in 2.0 (this is just a change in name)

jobara replied:
@amb26 similar to my other comment, would it be okay to change this in the work for FLOE-171 which will include an updated infusion.

jobara replied:
@amb26 made this change throughout

jobara committed:
FLOE-116: refactored conditionAny

Refactored to use fluid.find internally instead of a for loop.

jobara committed:
FLOE-116: renamed conditionAny

Renamed the conditionAny transform to find (fluid.transforms.find)

jobara committed:
FLOE-116: renamed init namespace and listener

Renamed to loadTemplate to be more descriptive.

jobara commented:
@amb26 I've made changes based on your comments. There were several things that you wanted me to change that would require an update to our instance of infusion. As I mentioned in the related replies , this branch is being used by one that @cindyli is working on to solve FLOE-171. In there she has already updated Infusion. I'm hoping we can fix these other issues there, so as to avoid potential conflicts and etc. Let me know what you think and if you have any other review comments.

amb26 commented on tests/js/resourceInputPanelTests.js:
```diff
@@ -0,0 +1,243 @@
+                    options: {
+                        listeners: {
+                            afterRender: {
+                                listener: function (input) {
```
This deeply nested literal renderer is badly expressed and should be written as a standalone global function

amb26 commented on tests/js/resourceInputPanelTests.js:
```diff
@@ -0,0 +1,243 @@
+
+        var styleClass = "fl-test-style";
+
+        var that = fluid.metadata.baseResourceInputPanel(".flc-baseResourceInputPanel", {
```
Try to write these definitions using grades and standalone global functions, rather than supplying all the configuration as a literal argument inside the test function. This will help to make your test definitions more reusable

jobara committed:
FLOE-116: Refactoring unit tests.

jobara commented:
@amb26 I refactored the tests a bit. I couldn't make all the changes though as there seems to be some bugs with the dynamicComponents in the version of Infusion I'm using. I made some further changes in @cindyli's FLOE-171 branch as the issues aren't present in the version of Infusion used there. Ready for another round of review.

michelled commented:
@colinbdclark, @amb26: would either of you be able to review the latest changes on this pull request?

jobara committed:
Merge branch 'master' into FLOE-116

Conflicts:
	demos/html/metadata.html
	src/css/style.css
	src/html/audio-template.html
	src/html/captions-input-template.html
	src/html/captions-template.html
	src/html/video-template.html
	src/js/audioPanel.js
	src/js/captionsPanel.js
	src/js/videoPanel.js
	tests/js/audioPanelTests.js
	tests/js/videoPanelTests.js

jobara committed:
FLOE-116: fixed include

jobara committed:
FLUID-116: using change instead of requestChange

jobara commented:
@amb26 ready for more code review

amb26 commented on src/js/panel.js:
```diff
@@ -0,0 +1,98 @@
+
+    // At every panel model change, transform it to the model value required by the indicator sub-component,
+    // as well as requesting the indicator to refresh its state. This model relay is only one way from the
+    // panel model to the indicator model.
```
Could you add a link here referring to http://issues.fluidproject.org/browse/FLUID-5337 and explain that when this is fixed, this manual inverse relay will be performed automatically by the framework

amb26 commented on src/js/panel.js:
```diff
@@ -0,0 +1,98 @@
+                        }
+                    },
+                    modelListeners: {
+                        "{panel}.model.*": {
```
This should read "{panel}.model" - please see our updated docs at http://wiki.fluidproject.org/display/docs/ChangeApplier+API for explanation of the rules

amb26 commented on tests/js/resourceInputPanelTestUtils.js:
```diff
@@ -0,0 +1,134 @@
+                    func: "fluid.tests.changeSrcByIndex",
+                    args: ["{resourceInputPanel}", "{that}.options.testOptions.newSrcValue1", 0]
+                }, {
+                    listenerMaker: "fluid.tests.checkModelValueByIndex",
```
This repetitive material should be factored into a block of configuration that then generates this sequence automatically. For example something like
[{check: "valueByIndex", path: "newSrcValue1"},
 {check: "languageByIndex", path: "newLanguage1", index: 0}
etc.

jobara committed:
FLOE-116: Updated comment

Added a comment for fluid.metadata.panel.updateModel, indicating that it will be taken care of by the framework when http://issues.fluidproject.org/browse/FLUID-5337 is addressed.

jobara committed:
FLOE-116: Updated modelListener path for panel.js

jobara committed:
FLOE-116: Left a comment about future updates

Left a comment about changing the modelListener path from "*" to "" after the components have been updated to the new model relay system. This work should be happening as part of the FLOE-171 changes.

jobara committed:
FLOE-116: Refactored tests

jobara commented:
@amb26 ready for another round of review.

jobara committed:
FLOE-116: refactored resourceInputPanel tests

Refactored the resourceInputPanel initialization test. This change had originally been done as part of the FLOE-171 work, but since master had been updated with the latest infusion I was able to move this change back into this branch. Also refactored a bit more to make use of a grade for the tested component.

jobara commented:
@amb26 i refactored the resourceInputPanel initialization tests.

jobara committed:
FLOE-116: Updated a comment
